src/aggregators/coevolve.jl

```diff
@@ -0,0 +1,206 @@
+end
+
+# set up a new simulation and calculate the first jump / jump time
+function initialize!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    p.end_time = integrator.sol.prob.tspan[2]
+    fill_rates_and_get_times!(p, u, params, t)
+    generate_jumps!(p, integrator, u, params, t)
+    nothing
+end
+
+# execute one jump, changing the system state
+function execute_jumps!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    # execute jump
+    u = update_state!(p, integrator, u, params, t)
+    # update current jump rates and times
+    update_dependent_rates!(p, u, params, t)
+    nothing
+end
+
+# calculate the next jump / jump time
+function generate_jumps!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    p.next_jump_time, p.next_jump = top_with_handle(p.pq)
+    nothing
+end
+
+@inline function update_state!(p::CoevolveJumpAggregation, integrator, u, params, t)
+    @unpack ma_jumps, next_jump = p
+    num_majumps = get_num_majumps(ma_jumps)
+    if next_jump <= num_majumps
+        if u isa SVector
+            integrator.u = executerx(u, next_jump, ma_jumps)
+        else
+            @inbounds executerx!(u, next_jump, ma_jumps)
+        end
+    else
+        idx = next_jump - num_majumps
+        @inbounds p.affects![idx](integrator)
+    end
+    p.prev_jump = next_jump
+    return integrator.u
+end
+
+######################## SSA specific helper routines ########################
+function update_dependent_rates!(p::CoevolveJumpAggregation, u, params, t)
+    @inbounds deps = p.dep_gr[p.next_jump]
+    @unpack end_time, pq = p
+    for (ix, i) in enumerate(deps)
+        ti = next_time(p, u, params, t, i, end_time)
+        update!(pq, i, ti)
+    end
+    nothing
+end
+
+function get_rates(p::CoevolveJumpAggregation, i, u)
+    ma_jumps = p.ma_jumps
+    num_majumps = get_num_majumps(ma_jumps)
+    if i <= num_majumps
+        _rate = evalrxrate(u, i, ma_jumps)
+        rate(u, p, t) = _rate
+        lrate = rate
+        urate = rate
+        L(u, p, t) = typemax(t)
+    else
+        idx = i - num_majumps
+        rate, lrate, urate, L = p.rates[idx], p.lrates[idx], p.urates[idx], p.Ls[idx]
+    end
+    return rate, lrate, urate, L
+end
+
+function next_time(p::CoevolveJumpAggregation{T}, u, params, t, i, tstop::T) where {T}
+    @unpack rng = p
+    rate, lrate, urate, L = get_rates(p, i, u)
+    while t < tstop
+        _urate = urate(u, params, t)
+        _L = L(u, params, t)
+        s = _urate == zero(t) ? typemax(t) : randexp(rng) / _urate
+        _t = t + s
+        if s > _L
+            t = t + _L
+            continue
+        end
+        if _t >= tstop
+            break
+        end
+        _lrate = lrate(u, params, t)
+        if _lrate > _urate
+            error("The lower bound should be lower than the upper bound rate for t = $(t) and i = $(i), but lower bound = $(_lrate) > upper bound = $(_urate)")
+        elseif _lrate < _urate
```

Comment:
So `lrate` is not strictly required for the algorithm; it is just helpful if for some problems it is easier to calculate than the full `rate` function?

In that case, couldn't we make it optional, and here if it's field is `nothing` just sample with the actual rate, but if it is present sample with both as you currently do? I guess we might want to have a bitvector that stores that information in the aggregator to avoid having to examine the type within the core loop here.

Reply:
That's right `lrate` is not strictly required. It just makes the algorithm faster if supplied. The safe default would be to have `lrate = 0` instead.

Reply:
We can make this a future performance tweak / update.

Reply:
I have already made `lrate` optional. So when `lrate` is not passed to `VariableRateJump` but `urate` and `L` are, `lrate` becomes: 
```julia
lrate = (u, p, t) -> zero(typeof(t))
```